src/ai/backend/manager/models/session.py

```diff
@@ -0,0 +1,1066 @@
+from .minilang.queryfilter import QueryFilterParser
+from .user import UserRow
+
+if TYPE_CHECKING:
+    from sqlalchemy.engine import Row
+
+    from .gql import GraphQueryContext
+
+
+__all__ = (
+    "aggregate_kernel_status",
+    "SessionStatus",
+    "DEAD_SESSION_STATUSES",
+    "AGENT_RESOURCE_OCCUPYING_SESSION_STATUSES",
+    "USER_RESOURCE_OCCUPYING_SESSION_STATUSES",
+    "SessionRow",
+    "SessionDependencyRow",
+    "check_all_dependencies",
+    "ComputeSession",
+    "ComputeSessionList",
+)
+
+
+class SessionStatus(enum.Enum):
+    # values are only meaningful inside the manager
+    PENDING = 0
+    # ---
+    SCHEDULED = 5
+    PREPARING = 10
+    # ---
+    RUNNING = 30
+    RESTARTING = 31
+    RUNNING_DEGRADED = 32
+    # ---
+    TERMINATING = 40
+    TERMINATED = 41
+    ERROR = 42
+    CANCELLED = 43
+
+
+DEAD_SESSION_STATUSES = (
+    SessionStatus.CANCELLED,
+    SessionStatus.TERMINATED,
+)
+
+# statuses to consider when calculating current resource usage
+AGENT_RESOURCE_OCCUPYING_SESSION_STATUSES = tuple(
+    e
+    for e in SessionStatus
+    if e
+    not in (
+        SessionStatus.TERMINATED,
+        SessionStatus.PENDING,
+        SessionStatus.CANCELLED,
+    )
+)
+
+USER_RESOURCE_OCCUPYING_SESSION_STATUSES = tuple(
+    e
+    for e in SessionStatus
+    if e
+    not in (
+        SessionStatus.TERMINATING,
+        SessionStatus.TERMINATED,
+        SessionStatus.PENDING,
+        SessionStatus.CANCELLED,
+    )
+)
+
+
+KERNEL_SESSION_STATUS_MAPPING: Mapping[KernelStatus, SessionStatus] = {
+    KernelStatus.PENDING: SessionStatus.PENDING,
+    KernelStatus.SCHEDULED: SessionStatus.SCHEDULED,
+    KernelStatus.PREPARING: SessionStatus.PREPARING,
+    KernelStatus.BUILDING: SessionStatus.PREPARING,
+    KernelStatus.PULLING: SessionStatus.PREPARING,
+    KernelStatus.RUNNING: SessionStatus.RUNNING,
+    KernelStatus.RESTARTING: SessionStatus.RUNNING,
+    KernelStatus.RESIZING: SessionStatus.RUNNING,
+    KernelStatus.SUSPENDED: SessionStatus.RUNNING,
+    KernelStatus.TERMINATING: SessionStatus.TERMINATING,
+    KernelStatus.TERMINATED: SessionStatus.TERMINATED,
+    KernelStatus.ERROR: SessionStatus.ERROR,
+    KernelStatus.CANCELLED: SessionStatus.CANCELLED,
+}
+
+
+def aggregate_kernel_status(
+    kernels: Sequence[KernelRow], target_status: KernelStatus
+) -> SessionStatus:
+    candidate: SessionStatus = KERNEL_SESSION_STATUS_MAPPING[target_status]
+    priority_statuses = [status for status in KernelStatus if status.value < target_status.value]
+    for s in kernels:
+        if s.status == KernelStatus.ERROR:
+            if s.cluster_role == DEFAULT_ROLE:
+                return SessionStatus.ERROR
+            return SessionStatus.RUNNING_DEGRADED
+        elif s.status in priority_statuses:
+            candidate = KERNEL_SESSION_STATUS_MAPPING[s]
+        elif s.status == target_status:
+            continue
+        else:
+            # TODO: handle unexpected kernel status
+            pass
```

Comment:
I think we can just raise `ValueError` instead of leaving this as `TODO`.

```suggestion
            raise ValueError(f'Unexpected status {s.status}')
```